bundles/io/org.eclipse.smarthome.io.security/src/main/java/org/eclipse/smarthome/io/security/api/SecretsVaultService.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.security.api;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The {@link SecretsVaultService} can be used to store and retrieve secrets.
+ *
+ * Secrets can only be retrieved by the bundle that stored them and by default
+ * will be encrypted and and stored on disk immediately.
+ *
+ * Because of the way how the String class is handled internally, you are strongly
+ * encouraged to use char or byte arrays or the ByteBuffer class
+ * for keeping encrypted secrets in memory.
+ *
+ * The encryption algorithm, the storage location and format depend on the implementation.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public interface SecretsVaultService {
+    /**
+     * Retrieve a secret, identified by the given key.
+     *
+     * @param key An identifier
+     * @return Either the secret as ByteBuffer that was stored by this bundle under the given key
+     *         or null if no secret could be found.
+     * @throws IOException If the underlying storage file can't be opened or read.
+     * @throws GeneralSecurityException If the decryption algorithm fails, this exception is thrown
+     */
+    @Nullable
+    ByteBuffer retrieve(String key) throws IOException, GeneralSecurityException;
+
+    /**
+     * Store a secret, identified by the given key.
+     *
+     * @param key An identifier
+     * @param secret A secret as a ByteBuffer. If the ByteBuffer is empty, an empty secret will be stored.
+     * @throws IOException If the underlying file can't be opened or is not writable.
+     * @throws GeneralSecurityException If the encryption algorithm fails, this exception is thrown
+     */
+    void add(String key, ByteBuffer secret) throws IOException, GeneralSecurityException;
```

Comment:
Should this method not throw an exception if the given key already exists and not silently overwrite it? Because as far as I understand the current API a modification has to go through `SecretsVaultManagement.modify()`?

Reply:
Yes. It should return a Boolean or throw.